docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md

```diff
@@ -0,0 +1,245 @@
+  - Enable/disable `no hardware accelerations` with configure options.
+- Without `runtime detection`( Build-time detection only).
+  - `runtime detection` for architecture MUST be disabled.
+  - If `no hardware accelerations` is enabled, all hardwares accelerations MUST be disabled.
+  - If `no hardware accelerations` is disabled, only one hardwares acceleration can be enabled.
+  - If possible, enable/disable hardware accelerations base on `CFLAGS`.
+
+NOTE: The algorithm can be implemented with different instruction sets on same architecture. For example, AES can be implemented with crypto extention or SVEAES extension in Arm64.
+
+
+## Suggestion
+
+### check config
+
+`check_config.h` MUST not include any archeticture special checks. This gives up the advantage of the current code. But that makes it architecture independent.
+
+For `build-time detection only`, we can enable the the checks in hardware acceleration module.
+
+There are a potential issue for `build-time detection only`, if library is built with some CPU modifer flags, how application known the parameter? Is it important ?
+
+### Configure options
+
+- `MBEDTLS_RUNTIME_DETECTION`: Enable runtime detection feature.
+- Target platform macros: `MBEDTLS_ARCH_IS_{X86,AMD64,ARM64}`
+  - All of them MUST be disabled and can be enabled from command line.
+  - Only one can be enabled.
+  - If not defined, define them base on cross-cc information
+- Hardware accelerations, it must follow bellow ruler.
+  - `MBEDTLS_<algo>_<arch>_<extension>`.
+  - AES example: `MBEDTLS_AES_ARM64_CRYPTO`,`MBEDTLS_AES_ARM64_SVEAES`,`MBEDTLS_AES_X86_PADLOCK`,`MBEDTLS_AES_AMD64_AES`
+- No hardware accelerations, follow bellow ruler.
+  - `MBEDTLS_<algo>_NO_ACCEL`.
+  - AES example: `MBEDTLS_AES_NO_ACCEL`
+
+#### Check config example
+
+```c
+#if defined(MBEDTLS_AES_ARM64_CRYPTO)
+#define MBEDTLS_AES_ARM64_CRYPTO_ENABLED 1
+#else
+#define MBEDTLS_AES_ARM64_CRYPTO_ENABLED 0
+#endif
+#if defined(MBEDTLS_AES_NO_ACCEL)
+#define MBEDTLS_AES_NO_ACCEL_ENABLED 1
+#else
+#define MBEDTLS_AES_NO_ACCEL_ENABLED 0
+#endif
+
+#if defined(MBEDTLS_RUNTIME_DETECTION)
+#if defined(MBEDTLS_ARCH_IS_ARM64) && ( MBEDTLS_AES_NO_ACCEL_ENABLED + \
+                                        MBEDTLS_AES_ARM64_CRYPTO_ENABLED ) == 0
+#error "At least one AES algo MUST be enabled"
+#endif
+#else
+#if defined(MBEDTLS_ARCH_IS_ARM64) && ( MBEDTLS_AES_NO_ACCEL_ENABLED +
+                                        MBEDTLS_AES_ARM64_CRYPTO_ENABLED ) != 1
+#error "Only and MUST enable one AES algo for buildtime detection"
+#endif
+#endif
+#undef MBEDTLS_AES_NO_ACCEL_ENABLED
+#undef MBEDTLS_AES_ARM64_CRYPTO_ENABLED
+```
+
+### Modules
+
+#### Runtime Detection
+
+This module depends on `MBEDTLS_RUNTIME_DETECTION`. It provides cpu feature detection function and feature masks.
+
+files:
+- `library/runtime.h`
+- `library/runtime.c`
+
+I prefer implement the function as util macro in `runtime.h` like below.
+
+```c
+#if defined(MBEDTLS_RUNTIME_DETECTION)
```

Comment:
Sure. I will add it as a sub option of `MBEDTLS_RUNTIME_C`

Reply:
Why do we need a custom detection function? What would be the use case?

Reply:
Now, we only support arm64/32 x86/x64. If user want to support risc-v , he can provide it.

Reply:
Arguably Risc-V is out of scope since we don't provide any Risc-V acceleration. If we ever add Risc-V acceleration, it should come with Risc-V feature detection.

But we need custom feature detection for arm! Unfortunately, on arm, feature detection requires the OS's collaboration. The instructions to query the CPU features rely on the OS. You can try an instruction and see if it faults, but catching the feault again relies on the OS. So we can provide methods for common OSes, but embedded OSes need to be able to provide their own.

Reply:
Oh I see, thank you both